libyul/backends/evm/EVMBuiltins.cpp

```diff
@@ -0,0 +1,418 @@
+{
+	return createFunction(
+		"auxdataloadn",
+		1,
+		1,
+		EVMBuiltins::sideEffectsOfInstruction(evmasm::Instruction::DATALOADN),
+		ControlFlowSideEffects::fromInstruction(evmasm::Instruction::DATALOADN),
+		{LiteralKind::Number},
+		[](
+			FunctionCall const& _call,
+			AbstractAssembly& _assembly,
+			BuiltinContext&
+		) {
+			yulAssert(_call.arguments.size() == 1);
+			Literal const* literal = std::get_if<Literal>(&_call.arguments.front());
+			yulAssert(literal, "");
+			yulAssert(literal->value.value() <= std::numeric_limits<uint16_t>::max());
+			_assembly.appendAuxDataLoadN(static_cast<uint16_t>(literal->value.value()));
+		}
+	);
+}
+
+BuiltinFunctionForEVM eofcreateBuiltin()
+{
+	return createFunction(
+		"eofcreate",
+		5,
+		1,
+		EVMBuiltins::sideEffectsOfInstruction(evmasm::Instruction::EOFCREATE),
+		ControlFlowSideEffects::fromInstruction(evmasm::Instruction::EOFCREATE),
+		{LiteralKind::String, std::nullopt, std::nullopt, std::nullopt, std::nullopt},
+		[](
+			FunctionCall const& _call,
+			AbstractAssembly& _assembly,
+			BuiltinContext& context
+		) {
+			yulAssert(_call.arguments.size() == 5);
+			Literal const* literal = std::get_if<Literal>(&_call.arguments.front());
+			auto const formattedLiteral = formatLiteral(*literal);
+			yulAssert(!util::contains(formattedLiteral, '.'));
+			auto const* containerID = util::valueOrNullptr(context.subIDs, formattedLiteral);
+			yulAssert(containerID != nullptr);
+			yulAssert(*containerID <= std::numeric_limits<AbstractAssembly::ContainerID>::max());
+			_assembly.appendEOFCreate(static_cast<AbstractAssembly::ContainerID>(*containerID));
+		}
+	);
+}
+
+BuiltinFunctionForEVM returncontractBuiltin()
+{
+	return createFunction(
+		"returncontract",
+		3,
+		0,
+		EVMBuiltins::sideEffectsOfInstruction(evmasm::Instruction::RETURNCONTRACT),
+		ControlFlowSideEffects::fromInstruction(evmasm::Instruction::RETURNCONTRACT),
+		{LiteralKind::String, std::nullopt, std::nullopt},
+		[](
+			FunctionCall const& _call,
+			AbstractAssembly& _assembly,
+			BuiltinContext& context
+		) {
+			yulAssert(_call.arguments.size() == 3);
+			Literal const* literal = std::get_if<Literal>(&_call.arguments.front());
+			yulAssert(literal);
+			auto const formattedLiteral = formatLiteral(*literal);
+			yulAssert(!util::contains(formattedLiteral, '.'));
+			auto const* containerID = util::valueOrNullptr(context.subIDs, formattedLiteral);
+			yulAssert(containerID != nullptr);
+			yulAssert(*containerID <= std::numeric_limits<AbstractAssembly::ContainerID>::max());
+			_assembly.appendReturnContract(static_cast<AbstractAssembly::ContainerID>(*containerID));
+		}
+	);
+}
+
+}
+
+EVMBuiltins::EVMBuiltins()
+{
+	for (auto const& [name, opcode]: evmasm::c_instructions)
+	{
+		if (
+			opcode == evmasm::Instruction::SWAPN ||
+			opcode == evmasm::Instruction::DUPN ||
+			evmasm::SemanticInformation::isSwapInstruction(opcode) ||
+			evmasm::SemanticInformation::isDupInstruction(opcode)
```

Comment:
These functions already check for `SWAPN`/`DUPN`.
```suggestion
			SemanticInformation::isSwapInstruction(opcode) ||
			SemanticInformation::isDupInstruction(opcode)
```

Reply:
Same in `isLowLevelStackManipulationInstruction()`.

Reply:
Also, shouldn't you be skipping `PUSH` and `isLowLevelControlFlowInstruction()` here as well? We don't expose them in any dialect.

Reply:
> These functions already check for `SWAPN`/`DUPN`.
> Same in `isLowLevelStackManipulationInstruction()`.

I purposefully did it that way because of how `AssemblyItem` is designed. We have the opcode here (type `evmasm::Instruction`) which is implicitly converted to `evmasm::AssemblyItem`. We [added asserts](https://github.com/ethereum/solidity/blob/78107e344ce526695a5a989250d01812a248d536/libevmasm/AssemblyItem.h#L89-L90) to not create assembly items for `SWAPN`/`DUPN` with the implicit conversion constructor but to always use the static helper methods. This makes it necessary to test for them individually and beforehand. It is quite unfortunate.

> Also, shouldn't you be skipping `PUSH` and `isLowLevelControlFlowInstruction()` here as well? We don't expose them in any dialect.

Not if we want the builtins thing to be a field over all possible (except for swap/dup) functions - which was my intention at least.